### PR TITLE
socket: return EPROTONOSUPPORT on invalid protocol

### PR DESCRIPTION
**Pull request by tiran**
Invalid socket protocol now fails with errno EPROTONOSUPPORT instead of
crashing with an assertion error.

### CODE REVIEW

File: src/library_sockfs.js [hunk 1]
```diff
@@ -48,8 +48,8 @@ mergeInto(LibraryManager.library, {
     createSocket: function(family, type, protocol) {
       type &= ~{{{ cDefine('SOCK_CLOEXEC') | cDefine('SOCK_NONBLOCK') }}}; // Some applications may pass it; it makes no sense for a single process.
       var streaming = type == {{{ cDefine('SOCK_STREAM') }}};
-      if (protocol) {
-        assert(streaming == (protocol == {{{ cDefine('IPPROTO_TCP') }}})); // if SOCK_STREAM, must be tcp
+      if (streaming && protocol && protocol != {{{ cDefine('IPPROTO_TCP') }}}) {
+        throw new FS.ErrnoError({{{ cDefine('EPROTONOSUPPORT') }}}); // if SOCK_STREAM, must be tcp or 0.
```
Comment (sbc100) [inline src/library_sockfs.js:52]: Is it really ok to be 0?

**Review by sbc100 — COMMENTED** [review]
Thanks!

File: tests/sockets/test_create_socket.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <assert.h>
+#include <errno.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (tiran) [reply tests/sockets/test_create_socket.c:18]: arpa/inet, signal, and string are not necessary.

File: src/library_sockfs.js (hunk 1 above)
Comment (tiran) [reply src/library_sockfs.js:52]: In my experience it is very common to use ``socket(AF_INET, SOCK_STREAM, 0)`` to create a IPv4 TCP socket. I have rarely seen ``IPPROTO_TCP``. It works fine on Linux, FreeBSD, macOS, and Windows.

Comment (sbc100) [reply src/library_sockfs.js:52]: Ah yes. I see.  From the man page: 

```
Normally only  a  single  protocol exists to support a particular socket type within a given
protocol family, in which case protocol can be specified as 0.
```

**Review by sbc100 — APPROVED** [review]

**Pull request merged**